lib/db_connection.ex

```diff
@@ -785,65 +894,382 @@ defmodule DBConnection do
+    default: `15_000`)
+    * `:log` - A function to log information about a call, either
+    a 1-arity fun, `{module, function, args}` with `DBConnection.LogEntry.t`
+    prepended to `args` or `nil`. See `DBConnection.LogEntry` (default: `nil`)
 
   The pool and connection module may support other options. All options
-  are passed to `handle_begin/2`, `handle_commit/2` and
-  `handle_rollback/2`.
+  are passed to `handle_begin/2`.
+
+  See `commit/2` and `rollback/2`.
 
   ### Example
 
-      {:ok, res} = DBConnection.transaction(conn, fn(conn) ->
-        DBConnection.execute!(conn, "SELECT id FROM table", [])
-      end)
+      {:ok, result} = DBConnection.begin(conn)
+      try do
+        DBConnection.execute(conn, "SELECT * FROM table", [])
+      after
+        DBConnection.commit(conn)
+      end
   """
-  @spec transaction(conn, (conn -> result), opts :: Keyword.t) ::
-    {:ok, result} | {:error, reason :: any} when result: var
-  def transaction(conn, fun, opts) do
-    case run(conn, &run_transaction/5, fun, meter(opts), opts) do
-      {:commit, {:ok, res, return, meter}} ->
-        log(meter, :transaction, :commit, nil, {:ok, res})
-        {:ok, return}
-      {:rollback, {:ok, res, reason, meter}} ->
-        log(meter, :transaction, :rollback, nil, {:ok, res})
-        {:error, reason}
-      {:rollback, {:ok, res, kind, reason, stack, meter}} ->
-        log(meter, :transaction, :rollback, nil, {:ok, res})
-        :erlang.raise(kind, reason, stack)
-      {query, {:error, err, meter}} when query != nil ->
-        log(meter, :transaction, query, nil, {:error, err})
+  @spec begin(conn, opts :: Keyword.t) :: {:ok, result} | {:error, Exception.t}
+  def begin(conn, opts \\ []) do
+    conn
+    |> run(&run_begin/4, meter(opts), opts)
+    |> log(:begin, :begin, nil)
+  end
+
+  @doc """
+  Execute a begin request.
+
+  Returns the `result` on success, otherwise raises an exception on error.
+
+  See `begin/2`.
+  """
+  @spec begin!(conn, opts :: Keyword.t) :: result
+  def begin!(conn, opts \\ []) do
+    case begin(conn, opts) do
+      {:ok, result} ->
+        result
+      {:error, err} ->
         raise err
-      {query, return} ->
-        log(return, :transaction, query, nil)
     end
   end
 
   @doc """
-  Rollback a transaction, does not return.
+  Acquire a lock on a connection and execute a begin request.
+
+  Return `{:ok, conn, result}` on success or `{:error, exception}` if there was
+  an error. If an error is returned the connection is released back to the pool.
+
+  The `conn` is a connection reference that can be used as the first argument to
+  other requests (except `checkout/2` and `checkout_begin/2`). If the connection
+  disconnects all future calls using that connection reference will return (or
+  raise if bang(!) variant) a `%DBConnection.ConnectionError{}` error.
+
+  Use `checkin/2`, `commit_checkin/2` or `rollback_checkin/2` to release the
```

Comment:
Nothing, but note we are going to have an option to force `rollback` on checkin (inside the pool).